plugin/plugin.go

```diff
@@ -363,6 +373,231 @@ func RegisterShutdown(name string, s Shutter) error {
 	return nil
 }
 
+// Configuration defines the interface for complex_configure callbacks, i.e. Go
+// structs that can hold and validate a configuration
+type Configuration interface {
+	Validate() error
+}
+
+var config_target *Configuration
+var config_targets map[string]*Configuration
+
+var configured map[string]chan struct{}
+
+// getOconfigChildren takes a C.oconfig_item_t and returns an array of Go
+// pointers to its child items. Getting the child pointers requires a little
+// more arithmetic than in C as we can't just increment a pointer to get the
+// next array item
+func getOconfigChildren(oconfig *C.oconfig_item_t) (output []*C.oconfig_item_t) {
+	start := unsafe.Pointer(oconfig.children)
+	size := unsafe.Sizeof(*oconfig.children)
+	for i := 0; i < int(oconfig.children_num); i++ {
+		child := (*C.oconfig_item_t)(unsafe.Pointer(uintptr(start) + size*uintptr(i)))
+		output = append(output, child)
+	}
+	return
+}
+
+// checkMatchable returns an error if the provided source cannot be copied to
+// the target, or nil if it's OK
+func checkMatchable(oconfig *C.oconfig_item_t, target reflect.Value) error {
```

Comment:
You're using a two step process: first you check if the config can be assigned, then you actually assign it. This duplicates all the checks and I don't understand if there is any benefit to it. Can't we simply call `assignConfig()` and document that if `RequestConfiguration()` returns an error that the provided struct may be in a half-initialized state and should not be used?